rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+Just as today, Kibana will shut the preboot HTTP server down as soon as it's ready to start the primary HTTP server.
+
+### 3.2.3 Elasticsearch service
+
+As mention in the [Motivation section](#2-motivation), the main goal of the interactive setup mode is to give the user a hassle-free way to configure Kibana connection to an Elasticsearch cluster. That means that users might provide certain connection information, and Kibana preboot plugins should be able to construct a new Elasticsearch client using this information to verify it and potentially call Elasticsearch APIs.
+
+To support this use case we'll add a new `preboot` method to the Elasticsearch service that will return the following contract, and make it available to the preboot plugins via `CorePrebootSetup`:
+
+```ts
+export interface ElasticsearchServicePrebootSetup {
+  readonly createClient: (
+    type: string,
+    clientConfig?: Partial<ElasticsearchClientConfig>
+  ) => ICustomClusterClient;
+}
+```
+
+The Elasticsearch clients created with `createClient` rely on the default Kibana Elasticsearch configuration and any configuration overrides specified by the consumer.
+
+__NOTE:__ We may need to expose a full or portion of Elasticsearch config to the preboot plugins for them to check if the user has already configured Elasticsearch connection. There are other ways to check that without direct access to the configuration though.
+
+### 3.2.4 UI Settings service
+
+We'll introduce a new `preboot` method in the UI Settings service that will produce a UI Settings client instance. Since during the `preboot` stage Kibana can access neither user information nor Saved Objects, this client will only give access to the **default** UI settings and their overrides set through Kibana configuration, if any:
+
+```ts
+export interface InternalUiSettingsServicePrebootSetup {
+  defaultsClient(): IUiSettingsClient;
+}
+```
+
+UI Settings service isn't strictly necessary during the `preboot` stage, but many Kibana Core components rely on it explicitly and implicitly, which justifies this simple change.
+
+### 3.2.5 Rendering service
+
+We'll introduce a new `preboot` method in the Rendering service that will register Kibana main UI bootstrap template route on the preboot HTTP server as it does for the primary HTTP server today. The main difference is that bootstrap UI will only reference bundles of the preboot plugins and will rely on the default UI settings.
+
+### 3.2.6 I18n service
+
+We'll introduce a new `preboot` method in the I18n service to only include translations for the Core itself and preboot plugins in the translations bundle loaded with the preboot UI bootstrap template. This would potentially allow us to switch locale during interactive setup mode if there is such a need in the future.
```

Comment:
>  This would potentially allow us to switch locale during interactive setup mode if there is such a need in the future.

Agree, that we shouldn't implement it from the very beginning. `Preboot` phase is used by admin, we don't translate any log messages for them.

Reply:
> `Preboot` phase is used by admin, we don't translate any log messages for them. 

We don't translate log messages, but we still localize the interactive setup mode plugin (I can see us or partners distributing Kibana binaries with the predefined non-English locale). But every byte counts, so preboot HTTP server will only be serving translations required by Core and preboot plugins.